[팀프로젝트][F][요기조기] 야놀자 클론코딩 1차 PR

sirin0762 opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->

## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
- F팀의 요기조기 숙박 예약 시스템(야놀자 클론코딩).
- 호텔 파트의 api를 구현했습니다.

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
### 구현

**지역 조회 페이지(https://www.yanolja.com/hotel)**

- 지역에 따른 호텔을 조회 할 수 있게 했습니다.


**예약 가능한 호텔 조회(https://www.yanolja.com/hotel/r-910161/?advert=AREA&hotel=1)**
- 예약 날짜와 인원에 따라 호텔을 보여줄수 있게 했습니다.
- 추가적으로 필터 기능을 통해서(호텔 성급, 테마) 필터링 할 수 있게 했습니다. 

**호텔 세부 조회(https://www.yanolja.com/hotel/3016868)**
- 호텔의 세부 정보를 볼 수 있습니다.
- 호텔의 전반적인 정보 및 이미지 url 를 제공하는 api, 리뷰를 제공하는 api, 예약가능한 룸을 제공하는 api 3가지가 포함되어 있습니다.
- 날짜 변경시, 예약가능한 방이 바뀌도록 구현했습니다. 단, 각 방의 재고는 1개로 고정되어있습니다.

**객실 상세 조회(https://www.yanolja.com/hotel/3016868/339462)**
- 객실의 세부 정보를 볼 수 있습니다.

**예약 기능**
- 사용자 정보와 방의 정보를 토대로, 예약 기능을 구현했습니다.

## ✅ 피드백 반영사항  <!-- 지난 코드리뷰에서 고친 사항을 적어주세요. 재PR 시에만 사용해 주세요! (재PR 아닌 경우 삭제) -->

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->

Commit:
FEAT - 1_all_create_entity : 엔티티 생성 및 연관관계 편의 메서드 구현

sirin0762 committed:
Merge pull request #1 from prgrms-be-devcourse/all_1_create_entity

FEAT - 1_all_create_entity : 엔티티 생성 및 연관관계 편의 메서드 구현

eden6187 committed:
FEAT : Entity Column 추가

Commit:
FEAT(teamf-12) : 호텔 단건 조회 메인페이지 데이터(이미지, 글) 뿌리기
- 호텔에 기본적으로 들어가는 정보(이름, 지역, 등급, 리뷰 수, 테마, 이미지 url) 을 가지고 올 수 있게 했습니다.
- 호텔에 이미지를 등록할 수 있게 했습니다.

Commit:
Remove(teamf-12) : 이미지 컨트롤러 및 테스트 삭제
- 이미지를 별도의 컨트롤러가 아닌 도메인에 종속되게 변경하였기 때문에 필요없어 삭제합니다.

eden6187 committed:
FEAT : 예약 가능한 호텔 지역 조회

Commit:
Test(teamf-12) : 테스트를 위한 설정파일 변경
-기존에 application.yaml 외에, aws.yaml 파일을통해서 따로 관리를 하려했으나, 테스트에서 이를 인식하지 못하여(어플리케이션은 인식함) 테스트 오류가 나타났습니다.
- 위의 2개의 파일틀 통합하여 aws 설정도 application.yaml 파일에 들어가게 했습니다.
- 또한 image 테스트의 경우는 source가 필요하기 때문에 application-test.yaml 파일을 통해서 테스트 하게 설정되있습니다.

Commit:
Refactor(teamf-12) : HotelService 생성방식 변경, ErrorMessage of 메서드 삭제

sirin0762 committed:
Merge pull request #2 from prgrms-be-devcourse/teamf-12

FEAT(teamf-12) : 호텔 단건 조회 메인페이지 데이터(이미지, 글) 뿌리기

eden6187 committed:
Merge branch 'develop' into TEAMF-15

eden6187 committed:
STYLE : ResponseEntity 로 Controller 응답 통일

eden6187 committed:
Merge pull request #3 from prgrms-be-devcourse/TEAMF-15

Teamf 15

Commit:
FEAT(TEAMF-13) : Hotel 생성 기능 구현

Commit:
FEAT(TEAMF-13) : 호텔 단건 조회 시, 2개의 리뷰를 응답으로 내려주는 api 구현, create 요청 객체에 대한 validation, GlobalExceptionHandler 추가

Refactor : 특정객체에 의존하는 엔티티의 연관관계 매핑 방식을 바꿨습니다(해당 부분은 알아두셔야하니 코드에 설명 드릴게요)

Chaeonida committed:
사용자는 체크인과 체크아웃 날짜를 통해서 해당 객실이 숙박이 가능한지 조회

Commit:
Refactor : 특정 연관관계 편의 메서드를 객체를 생성할 떄 실행되도록 변경했습니다.

- room은 생성할 때 hotel을
- reservation은 생성할 때 user와 room을
- review는 생성할 때, hotel과 reservation을 가지고 있어야합니다

test : 연관관계 관련 테스트를 수정했습니다.

Chaeonida committed:
Test. 예외관련 테스트 추가 , 룸테스트 오류 수정

Chaeonida committed:
Merge remote-tracking branch 'origin/develop' into feature_8

# Conflicts:
#	src/main/java/com/programmers/yogijogi/entity/Hotel.java
#	src/main/java/com/programmers/yogijogi/entity/Reservation.java
#	src/main/java/com/programmers/yogijogi/entity/Room.java
#	src/main/java/com/programmers/yogijogi/repository/HotelRepository.java
#	src/main/java/com/programmers/yogijogi/repository/RoomRepository.java
#	src/main/java/com/programmers/yogijogi/service/HotelService.java

Chaeonida committed:
Refactor. 룸과 호텔 api 경로가 겹쳐서 수정

Chaeonida committed:
Merge pull request #5 from prgrms-be-devcourse/feature_8

Teamf 8 호텔 객실 조회

eden6187 committed:
FEAT : 체크인/아웃 날짜, 투숙 인원, Province(지역)을 이용한 예약 가능한 호텔 정보 조회

Commit:
Merge remote-tracking branch 'origin/develop' into TEAMF-13

# Conflicts:
#	src/main/java/com/programmers/yogijogi/entity/Room.java
#	src/test/java/com/programmers/yogijogi/entity/RoomTest.java

Commit:
Test : teamf-8과 teamf-13 의 병합 중 문제가 생긴 테스트 케이스를 고쳤습니다.

sirin0762 committed:
Merge pull request #4 from prgrms-be-devcourse/TEAMF-13

Teamf 13 : 평점 및 리뷰는 별도의 API로 내려준다.

Commit:
Merge remote-tracking branch 'origin/TEAMF-3' into TEAMF-16

# Conflicts:
#	src/main/java/com/programmers/yogijogi/controller/HotelApiController.java
#	src/main/java/com/programmers/yogijogi/entity/Hotel.java
#	src/main/java/com/programmers/yogijogi/entity/Room.java
#	src/main/java/com/programmers/yogijogi/entity/Theme.java
#	src/main/java/com/programmers/yogijogi/repository/RoomRepository.java
#	src/main/java/com/programmers/yogijogi/service/HotelService.java
#	src/test/java/com/programmers/yogijogi/controller/HotelApiControllerTest.java

Commit:
TEAMF-16 : TEAMF-3과 develop의 병합 코드입니다.

- 기존에 객체 생성시 바로 연관관계를 매핑하는 방식을 연관관계 편의 메서드를 통해서 매핑하도록 변경되었습니다.

Chaeonida committed:
Feat. 객실 상세 조회 추가

Chaeonida committed:
Refactor. 예외처리 방식 변경

Chaeonida committed:
Refactor. 테스트 했을때 오류 수정

Chaeonida committed:
Merge pull request #8 from prgrms-be-devcourse/TEAMF-17

Teamf 17 : 객실 상세 페이지 추가

Commit:
Merge remote-tracking branch 'origin/develop' into TEAMF-16

sirin0762 committed:
Merge pull request #7 from prgrms-be-devcourse/TEAMF-16

Teamf 16 : TEAMF-3 과 develop 병합

Commit:
FEAT : TEAMF-18, TEAMF-19

- 호텔 단건 조회시, 날짜를 반영하여 룸을 보여주는 기능으로 추가됬습니다.
- 룸 이미지를 추가하는 기능을 추가했습니다.

Chaeonida committed:
Feat. 예약하는 기능 추가

sirin0762 committed:
Merge pull request #9 from prgrms-be-devcourse/TEAMF-16

FEAT : TEAMF-18, TEAMF-19

Chaeonida committed:
Feat. userId 통해 예약조회 기능 추가

Commit:
Refactor : TEAMF-20

- 전반적인 코드 리팩토링
- import 문 정리
- TEAMF-10과 MERGE

sirin0762 committed:
Merge pull request #11 from prgrms-be-devcourse/TEAMF-20

Refactor : TEAMF-20

sirin0762 committed:
Update README.md

lleellee0 commented on src/test/java/com/programmers/yogijogi/service/HotelServiceTest.java:
```diff
@@ -0,0 +1,246 @@
+                          Room.builder().maxGuest(TEST_GUEST_CNT).build()
+                              .addReservations(List.of(
+                                  Reservation.builder().checkIn(TEST_BASE_DATE)
+                                      .checkOut(TEST_BASE_DATE.plusDays(1)).build()
```
@sirin0762 @Chaeonida 

이렇게 depth가 너무 깊게 들어가면 코드 가독성이 아주 떨어집니다.
이정도로 depth가 길어진다면, 미리 변수에 담아서 파라미터로 넘겨주는게 더 가독성이 높은 코드가 됩니다.

sirin0762 replied:
depth 가 길어진다면, 별도의 함수 or 변수로 빼라는 뜻으로 이해했습니다!!
반영해보겠습니다 감사합니다😁

lleellee0 commented on src/main/java/com/programmers/yogijogi/service/RoomService.java:
```diff
@@ -0,0 +1,98 @@
+            List<Reservation> reservations = reservationRepository.getAllByRoom(room);
+            for (Reservation reservation : reservations) {
+                if ((checkIn.isBefore(reservation.getCheckOut()) && checkIn.isAfter(reservation.getCheckIn())) ||
+                        (checkOut.isBefore(reservation.getCheckOut()) && checkOut.isAfter(reservation.getCheckIn()))) {
```
@sirin0762 @Chaeonida 

조금 어려운 얘기가 될 수도 있는데, reservation의 CheckOut, CheckIn을 가져와서 지금 이 서비스에서 비교하는게 좋을까요, 아니면 reservation에게 checkIn과 checkOut을 던져서 reservation 내부에서 비교되도록 하는게 좋을까요?
한번 고민해보시기 바랍니다.

sirin0762 replied:
맞습니다. 저희도 처음 이야기를 했을 때, DB에서 비교가 일어나서 서비스는 비교된 객체들만 가져오게 하려 했지만 실력부족으로 하지 못했습니다 😥 좀더 공부해서 반영해보도록 하겠습니다!

lleellee0 commented on src/main/java/com/programmers/yogijogi/service/RoomService.java:
```diff
@@ -0,0 +1,98 @@
+
+
+    @Transactional
+    public List<RoomDetailDto> findAllByDate2(Long hotelId, LocalDate checkIn, LocalDate checkOut) throws NotFoundException {
```
@sirin0762 @Chaeonida 

이게.. 좋은 메서드 이름일까요?
더 적합한 메서드 이름이 있을 것 같습니다.

lleellee0 commented on src/main/java/com/programmers/yogijogi/service/HotelService.java:
```diff
@@ -0,0 +1,157 @@
+                (reservation.getCheckIn().isEqual(checkIn)
+                        || reservation.getCheckIn().isAfter(checkIn))
+                        && (reservation.getCheckIn().isEqual(checkOut)
+                        || reservation.getCheckIn().isBefore(checkOut));
```
@sirin0762 @Chaeonida 

제 생각에는 위 코드는 아래처럼 묶이는게 더 가독성이 높아질 것 같습니다.

```
(reservation.getCheckIn().isEqual(checkIn) || reservation.getCheckIn().isAfter(checkIn))
  && (reservation.getCheckIn().isEqual(checkOut) || reservation.getCheckIn().isBefore(checkOut));
```

근데 위 내용이랑 별개로, reservation의 checkIn, checkOut을 꺼내오지말고 reservation 내부에서 처리하도록 바꿔보세요.
https://github.com/prgrms-be-devcourse/BEVD1_F_YogiJogi/pull/12/files#r742692664
위 내용과 같은 내용입니다.

sirin0762 replied:
넵넵!! 한번 해보겠습니다!😁

lleellee0 commented on src/main/java/com/programmers/yogijogi/service/HotelService.java:
```diff
@@ -0,0 +1,157 @@
+                (reservation.getCheckOut().isEqual(checkIn)
+                        || reservation.getCheckOut().isAfter(checkIn))
+                        && (reservation.getCheckOut().isEqual(checkOut)
+                        || reservation.getCheckOut().isBefore(checkOut));
```
@sirin0762 @Chaeonida 

이것도 마찬가지입니다.

lleellee0 commented on src/main/java/com/programmers/yogijogi/service/RoomService.java:
```diff
@@ -0,0 +1,98 @@
+            for (Reservation reservation : reservations) {
+                if ((checkIn.isBefore(reservation.getCheckOut()) && checkIn.isAfter(reservation.getCheckIn())) ||
+                        (checkOut.isBefore(reservation.getCheckOut()) && checkOut.isAfter(reservation.getCheckIn()))) {
+                    reservation.getRoom().minusRoomStock();
```
@sirin0762 @Chaeonida 

여기 있는 `reservation.getRoom().minusRoomStock();` 역시 reservation 안으로 들어갔다면 훨씬 응집력있는 코드가 되었을 것 같습니다.
Reservation과 직접적으로 관련있는 코드라고 생각됩니다. (비즈니스 로직)

lleellee0 commented on src/main/java/com/programmers/yogijogi/entity/dto/ReservableHotelResponseDto.java:
```diff
@@ -0,0 +1,17 @@
+  Double reviewAverage;
+  Integer reviewCnt;
+  Integer price;
+  List<Image> images;
```
@sirin0762 @Chaeonida 

어떤 곳에서는 들여쓰기가 스페이스 2개,
또 다른 곳에서는 들여쓰기가 스페이스 4개로 되어있습니다.
통일시켜주는게 좋을 것 같습니다.

sirin0762 replied:
팀원 들의 인텔리제이 설정이 띄어쓰기 2칸, 4칸으로 각각이라서 해당 부분에 통일성이 떨어졌습니다 😥
코드 병합할 때 좀 더 유의깊게 살펴보겠습니다!

Commit:
Refactor : TEAMF-21

- ReservableHotelResponseDto 를 프론트엔드와 맞는 dto로 변경
- room Image를 1:다 로 변경
- 프론트를 위해서 CorsConfig 추가

sirin0762 committed:
Merge pull request #13 from prgrms-be-devcourse/TEAMF-20

Refactor : TEAMF-21

sirin0762 committed:
Update README.md

Chaeonida committed:
Feat. User 회원가입, 조회, 회원정보 수정 기능 추가

Chaeonida committed:
Refactor. /users 로 url 변경

Chaeonida committed:
Merge pull request #14 from prgrms-be-devcourse/TEAMF-22

Feat. User 회원가입, 조회, 회원정보 수정 기능 추가

Chaeonida committed:
Feat. Reservation 삭제 기능  추가 (soft delete)

Chaeonida committed:
Feat. Reservation 삭제 기능  추가 (soft delete)

Chaeonida committed:
Refactor. 주석 수정

Chaeonida committed:
Delete ReservationApiControllerTest.java

Chaeonida committed:
Revert "Delete ReservationApiControllerTest.java"

This reverts commit 9754ead1eb8de14f5747929d7cbef8d415ea8424.

Chaeonida committed:
Revert "Refactor. 주석 수정"

This reverts commit 52e6e026bb472c51575150ef6eedab290ce12a86.

Chaeonida committed:
Revert "Feat. Reservation 삭제 기능  추가 (soft delete)"

This reverts commit 282f49c0b4983e29fe14e9a2b644e7a4c1021a30.

Chaeonida committed:
Refactor. 주석 수정

Chaeonida committed:
Refactor. 주석 수정

Chaeonida committed:
Merge pull request #15 from prgrms-be-devcourse/TEAMF-23

Feat. Reservation 삭제 기능  추가 (soft delete)

lleellee0 commented on src/main/java/com/programmers/yogijogi/common/S3Uploader.java:
```diff
@@ -0,0 +1,72 @@
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class S3Uploader {
```
@sirin0762 @Chaeonida 

이 클래스 전체에 대해 한번 생각해볼만한 포인트가 있습니다.
객체지향에 대해 설명하라면 뭐라고 설명하실건가요?
초심자들이 생각하는 답변은 아마 여러가지 일겁니다.
'실제 세계에 있는 것을 모방하여 객체로 만드는 ....' 같은 내용을 이야기 하시려고 하겠죠?

하지만 틀렸습니다. **객체지향의 핵심은 바로 '추상화'**입니다.
그런데... 추상화가 핵심이라니. 정말 추상적인 내용이죠?

추상화를 좀 더 쉽게 설명하면, 추상적인 존재에 의존하도록 만들라는 것입니다.
S3Uploader를 생각해봅시다. 이 클래스는 추상적인 존재인가요?
아닙니다. 매우 구체적인 존재입니다.
S3Uploader를 사용하려는 클라이언트들은 모두 S3Uploader라는 구체적인 존재에 의존하게 됩니다.

그럼 추상적인 존재로 바꾸려면 어떻게 해야할까요?
그 답은 바로 해당 역할을 하는 인터페이스를 선언하는겁니다. 인터페이스의 이름은 Uploader 라고 해보죠.
기존에 S3Uploader를 사용하던 클라이언트들은 모두 S3Uploader가 아니라, Uploader라는 인터페이스에 의존하도록 바꿉니다.
그리고 S3Uploader는 Uploader 인터페이스를 구현하는 구현체로 만들면 되겠죠.

왜 이렇게 만들어야할까요?
그 이유는 구체적인 존재는 쉽게 변경되기 때문입니다.
만약 기존 코드대로 작성했는데, 우리가 AWS S3 대신 더 저렴한 파일 서비스를 사용하기로 했다면 어떻게할건가요?
혹은 그 둘다 사용하기로 했다면? 전자의 경우 S3라고 되어있는 모든 코드를 변경해주거나, 후자의 경우  새로운 구체 클래스를 만들어줘야 할겁니다. 그리고 그 인터페이스는 시스템이 운영되고 유지보수 되면서 서로 다른 인터페이스(서로 다른 public 메서드 집합을 가진다는 겁니다.)를 가지게 될지도 모릅니다.

lleellee0 commented on src/main/java/com/programmers/yogijogi/config/CorsConfig.java:
```diff
@@ -0,0 +1,15 @@
+    @Override
+    public void addCorsMappings(CorsRegistry registry) {
+        registry.addMapping("/**")
+                .allowedOrigins("*");
```
@sirin0762 @Chaeonida 

원래대로라면 허용된 Origin에 대해서만 허용해줘야합니다.
나중에 실제 서비스를 운영한다면 잘 따져보시기 바랍니다. :)

lleellee0 commented on src/main/java/com/programmers/yogijogi/controller/HotelApiController.java:
```diff
@@ -0,0 +1,117 @@
+            @PathVariable("id") Long hotelId) throws IOException {
+        String url = s3Uploader.upload(multipartFile, HOTEL_DIRNAME + "/" + hotelId);
+        hotelService.saveHotelImageUrl(hotelId, url);
+        return ResponseEntity.ok(url);
```
@sirin0762 @Chaeonida 

MultipartFile을 서비스로 가져가고 싶지 않아서 이렇게 만드신 것 같은데, 글쎄요...
차라리 퍼사드 서비스를 추가해서 해당 서비스가 파일업로드 + HotelService 호출까지 하면 어떨까 싶네요.
물론 너무 복잡해지는 것 같다고 생각되면 지금 구조도 괜찮습니다.

lleellee0 commented on src/main/java/com/programmers/yogijogi/service/UserService.java:
```diff
@@ -0,0 +1,73 @@
+    public UserService(UserRepository userRepository, UserConverter userConverter) {
+        this.userRepository = userRepository;
+        this.userConverter = userConverter;
+    }
```
@sirin0762 @Chaeonida 

보통 생성자보다 필드가 먼저오도록 코드를 작성합니다.
순간 제 눈을 의심했네요. ㅎ.. (저도 모르게 자바 문법이 바뀌어서 필드 선언 없이 사용이 되나해서)

lleellee0 commented on src/main/java/com/programmers/yogijogi/entity/dto/ReservableHotelRequestDto.java:
```diff
@@ -0,0 +1,35 @@
+public class ReservableHotelRequestDto {
+
+  // 아래의 4가지 요소는 필수적인 필드입니다.
+  // 4가지 필드를 받지 못한 이상 의미가 없어지기 때문에 빌더를 사용하지 않았습니다.
```
@sirin0762 @Chaeonida 

spring-boot-starter-validation 을 사용하는게 더 좋지 않을까.. 의견 드려봅니다.
한번 찾아보시고 선택해보세요.

lleellee0 replied:
다른 dto 들도 마찬가지로 유효성 검사를 dto에서 응집도 높게 처리하는걸 고려해보시기 바랍니다.

lleellee0 commented on src/main/java/com/programmers/yogijogi/entity/Host.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.ArrayList;
+import java.util.List;
+
+//많이쓰이는쪽에 add 반대쪽은 set
```
@sirin0762 @Chaeonida 

이 주석은 무슨 의미죵..?
